send ccip-read post request with json payload

mdtanrikulu opened the pull request:
### What was wrong?

ccip-read post request was failing due to, python `requests` package sends data as `application/x-www-form-urlencoded` when payload sent via `data=` flag.

### How was it fixed?

This PR instead sent the request payload as `application/json` content type.

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

:dog:

mdtanrikulu committed:
bugfix: CCIP-Read POST should send json data:

- send ccip-read post request with json payload
- update _mock_specific_request util for ccip-read post requests
- lint, remove additional header since it's added when payload sent with json arg

fselmo commented:
Thanks @mdtanrikulu. Do you have an example with an actual `OffchainResolver` we can use to test?

mdtanrikulu commented:
heya @fselmo 👋  this is the exact example I was testing against to the last changes 👇 

```python
from web3 import Web3, HTTPProvider
w3 = Web3(HTTPProvider("https://cloudflare-eth.com/"))

w3.ens.address("gregskril.com")
```

it hits to the `https://dnssec-oracle.ens.domains/` gateway with following payload 👇 

```json
{
  "sender": "0xf142b308cf687d4358410a4cb885513b30a42025",
  "data": "0x31b137b900000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000f0967726567736b72696c03636f6d000000000000000000000000000000000000"
}
```

which revealed the need for this change. fails if payload is data, passes if json.

fselmo commented:
> heya @fselmo 👋 this is the exact example I was testing against to the last changes 👇
> 
> ```python
> from web3 import Web3, HTTPProvider
> w3 = Web3(HTTPProvider("https://cloudflare-eth.com/"))
> 
> w3.ens.address("gregskril.com")
> ```
> 
> it hits to the `https://dnssec-oracle.ens.domains/` gateway with following payload 👇
> 
> ```json
> {
>   "sender": "0xf142b308cf687d4358410a4cb885513b30a42025",
>   "data": "0x31b137b900000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000f0967726567736b72696c03636f6d000000000000000000000000000000000000"
> }
> ```
> 
> which revealed the need for this change. fails if payload is data, passes if json.

Perfect I did test it and I can see the difference. I am going to make sure we can tighten up our tests around this first but this change looks good to me 👍🏼

fselmo commented:
@mdtanrikulu I am going to add some commits in here. Are you finished adding what you wanted to add? 👀

mdtanrikulu commented:
> @mdtanrikulu I am going to add some commits in here. Are you finished adding what you wanted to add? 👀

yep, it's done. feel free!

fselmo committed:
Add async changes for previous commit:

- Use json payload for CCIP-Read POST requests.
- Tweak mocker to check for ``json`` key and data in kwargs.
- Add newsfragment for PR #3512.

fselmo commented:
Thanks for the update here @mdtanrikulu 👍🏼. I added the async changes and a newsfragment. Tested locally as well and since the tests check the `json` field I think we're good to go here.
